### PR TITLE
docs: Update the CudaPackage (build system) description

### PR DESCRIPTION
**Pull request by tldahlgren**
This PR expands on the `CudaPackage` description by:

- elaborating on the variants;
- adding relevant URLs/links;
- adding a discussion of conflicts with examples;
- adding a description of the `cuda_flags` method; and
- correcting and expanding on the `Usage` section.

### CODE REVIEW

File: lib/spack/docs/build_systems/cudapackage.rst [hunk 1]
```diff
@@ -9,35 +9,120 @@
 CudaPackage
 -----------
 
-Different from other packages, ``CudaPackage`` does not represent a build
-system. Instead its goal is to simplify and unify usage of ``CUDA`` in other
-packages.
+Different from other packages, ``CudaPackage`` does not represent a build system.
+Instead its goal is to simplify and unify usage of ``CUDA`` in other packages.
```
Comment (ax3l) [inline lib/spack/docs/build_systems/cudapackage.rst]:
```suggestion
Instead its goal is to simplify and unify usage of ``CUDA`` in other packages by providing a ` mixin-class <https://en.wikipedia.org/wiki/Mixin>`__.
```

File: lib/spack/docs/build_systems/cudapackage.rst [hunk 2]
```diff
@@ -9,35 +9,120 @@
...
+* **cuda_arch**
+
+  This variant supports the optional specification of the architecture.
+  Valid values are maintained in the ``cuda_arch_values`` property and
+  are the numeric character equivalent of the compute capability version
+  (e.g., '10' for version 1.0). Each provided value affects associated
+  ``CUDA`` dependencies and compiler conflicts.
+
+  GPUs and their compute capability versions are listed at
+  https://en.wikipedia.org/wiki/CUDA#GPUs_supported.
```
Comment (ax3l) [inline lib/spack/docs/build_systems/cudapackage.rst]: I think the official, long-term page for this is https://developer.nvidia.com/cuda-gpus
```suggestion
  https://developer.nvidia.com/cuda-gpus .
```

File: lib/spack/docs/build_systems/cudapackage.rst [hunk 3]
```diff
@@ -9,35 +9,120 @@
...
+    conflicts('cuda_arch=none', when='+cuda',
+              msg='CUDA architecture is required')
+
+Similarly, if your software does not support all versions of the property,
+you could add ``conflicts`` to your package for those versions.  For example,
+suppose your software does not work with CUDA compute capability versions
+prior to `5.0`.  You can add the following code to display a custom message
+should a user attempt such a build:
+
+.. code-block:: python
```
Comment (ax3l) [inline lib/spack/docs/build_systems/cudapackage.rst:63]: This is a great example and I think we could even implement a helper method in a follow-up to pass a minimum numeric SM value, reducing boiler-plate in packages.

File: lib/spack/docs/build_systems/cudapackage.rst [hunk 4]
```diff
@@ -9,35 +9,120 @@
...
+
+    conflicts('cuda_arch=none', when='+cuda',
+              msg='CUDA architecture is required')
+
+Similarly, if your software does not support all versions of the property,
+you could add ``conflicts`` to your package for those versions.  For example,
+suppose your software does not work with CUDA compute capability versions
+prior to `5.0`.  You can add the following code to display a custom message
+should a user attempt such a build:
```
Comment (ax3l) [inline lib/spack/docs/build_systems/cudapackage.rst]: Just for clarification with dots and w/o:
```suggestion
prior to SM 5.0 (``50``).  You can add the following code to display a
custom message should a user attempt such a build:
```

**Review by ax3l — COMMENTED** [review]
Thank you for improving the docs of this! :sparkles:

**Comment by adamjstewart** [comment]
@Rombur would be a good person to review too

File: lib/spack/docs/build_systems/cudapackage.rst [hunk 5]
```diff
@@ -9,35 +9,120 @@
...
             if '+cuda' in spec:
-                options.append('-DWITH_CUDA=ON')
-                cuda_arch = spec.variants['cuda_arch'].value
+                # Set up the cuda macros needed by the build
+                args.append('-DWITH_CUDA=ON')
+                cuda_arch_list = spec.variants['cuda_arch'].value
+                cuda_arch = cuda_arch_list[0]
                 if cuda_arch != 'none':
-                    options.append('-DCUDA_FLAGS=-arch=sm_{0}'.format(cuda_arch[0]))
+                    args.append('-DCUDA_FLAGS=-arch=sm_{0}'.format(cuda_arch))
```
Comment (adamjstewart) [inline lib/spack/docs/build_systems/cudapackage.rst:113]: Should we be using `cuda_flags` here? The base class has a static method for this.

File: lib/spack/docs/build_systems/cudapackage.rst (hunk 3 above)
Comment (alalazo) [reply lib/spack/docs/build_systems/cudapackage.rst:63]: I think we can do that using metaclasses and requesting packages to just provide a map:
```
min_cuda_arch = [
    (min_value, when_spec_str),
    (min_value, when_spec_str),
    ...
]
```
Would that be a good enough interface?

Comment (ax3l) [reply lib/spack/docs/build_systems/cudapackage.rst:63]: absolutely, great idea!

Comment (tldahlgren) [reply lib/spack/docs/build_systems/cudapackage.rst:63]: Resolving this comment under the assumption the work discussed will be addressed in a follow-on PR for the metaclass and map approach suggested by alalazo.

File: lib/spack/docs/build_systems/cudapackage.rst (hunk 5 above)
Comment (tldahlgren) [reply lib/spack/docs/build_systems/cudapackage.rst:113]: The base class' `cuda_flags` returns a list of `'--generate-code arch=compute_{0},code=sm_{0} --generate-code arch=compute_{0},code=compute_{0}'.format(s)` for `s` in the list of arch's passed it.  Is that equivalent when the flags are specified this way?

Comment (tldahlgren) [reply lib/spack/docs/build_systems/cudapackage.rst:113]: Note the `dealii` package has a comment that it has compiler errors when using the values returned from `cuda_flags` (see https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py).

Comment (tldahlgren) [reply lib/spack/docs/build_systems/cudapackage.rst:113]: The `dihydrogen` package only uses `-arch=sm_60` (for its `CMAKE_CUDA_FLAGS` macro) when there is a single architecture with the value `auto`.

**Comment by tldahlgren** [comment]
@Rombur @adamjstewart Are these changes sufficient under the circumstances?

**Comment by adamjstewart** [comment]
Docs look good to me. I know very little about CUDA, I just use it for PyTorch. One thing I might suggest adding is an example where you determine the `cuda_arch` for your GPU and set it globally in `packages.yaml`.

As for the `CudaPackage` base class, I've been thinking this for a while now, but doesn't it make more sense to move the compiler-/platform-specific conflicts to the `cuda` package? That way they will always be used even if a package doesn't inherit from `CudaPackage`.

**Review by Rombur — APPROVED** [review]

**Comment by tldahlgren** [comment]
> Docs look good to me. I know very little about CUDA, I just use it for PyTorch. One thing I might suggest adding is an example where you determine the `cuda_arch` for your GPU and set it globally in `packages.yaml`.

As I mentioned in the new ROCmPackage doc, I'm hesitant to add something like this to here since there is a risk it could cause issues in some environments.  In that case, I referred to comments in the package (with a link to the package).

Unfortunately, the CudaPackage doesn't have comments with that advice.

> 
> As for the `CudaPackage` base class, I've been thinking this for a while now, but doesn't it make more sense to move the compiler-/platform-specific conflicts to the `cuda` package? That way they will always be used even if a package doesn't inherit from `CudaPackage`.

I'm not sure I follow the issue here.  What do you mean by the `cuda` package?  The `cuda.py` (build system) defines `CudaPackage`.

**Comment by adamjstewart** [comment]
> What do you mean by the cuda package?

`var/spack/repos/builtin/packages/cuda/package.py`

`CudaPackage` is just a set of common mixins for CUDA things. But not every package that uses CUDA uses `CudaPackage` or needs the variants. If we move this stuff to the package recipe, it will always be used for every package.

**Comment by tldahlgren** [comment]
> > What do you mean by the cuda package?
> 
> `var/spack/repos/builtin/packages/cuda/package.py`

Ah!  Guess my head is buried too much in the core.

> 
> `CudaPackage` is just a set of common mixins for CUDA things. But not every package that uses CUDA uses `CudaPackage` or needs the variants. If we move this stuff to the package recipe, it will always be used for every package.

Thanks for the clarification.

**Comment by tldahlgren** [comment]
@adamjstewart Since Rombur approved this PR, is the content sufficient for now under the assumption it can be updated, as needed, should any of the changes you mentioned be performed?

**Review by adamjstewart — APPROVED** [review]

**Comment by tldahlgren** [comment]
Thanks @adamjstewart

**Pull request merged**